two_factor_auth.go

```diff
@@ -0,0 +1,56 @@
+package packngo
+
+const twoFactorAuthAppPath = "/user/otp/app"
+const twoFactorAuthSmsPath = "/user/otp/sms"
+
+// TwoFactorAuthService interface defines available two factor authentication functions
+type TwoFactorAuthService interface {
+	EnableApp(string) (*Response, error)
+	DisableApp(string) (*Response, error)
+	EnableSms(string) (*Response, error)
+	DisableSms(string) (*Response, error)
+	ReceiveSms() (*Response, error)
+	ReceiveApp() (string, *Response, error)
+}
+
+// TwoFactorAuthServiceOp implements TwoFactorAuthService
+type TwoFactorAuthServiceOp struct {
+	client *Client
+}
+
+// EnableApp function enables two factor auth using authenticatior app
+func (s *TwoFactorAuthServiceOp) EnableApp(token string) (resp *Response, err error) {
+	headers := map[string]string{"x-otp-token": token}
+	return s.client.DoRequestWithHeader("POST", headers, twoFactorAuthAppPath, nil, nil)
+}
+
+// EnableSms function enables two factor auth using sms
+func (s *TwoFactorAuthServiceOp) EnableSms(token string) (resp *Response, err error) {
+	headers := map[string]string{"x-otp-token": token}
+	return s.client.DoRequestWithHeader("POST", headers, twoFactorAuthSmsPath, nil, nil)
+}
+
+// ReceiveSms orders the auth service to issue an SMS token
+func (s *TwoFactorAuthServiceOp) ReceiveSms() (resp *Response, err error) {
+	return s.client.DoRequest("POST", twoFactorAuthSmsPath+"/receive", nil, nil)
+}
+
+// DisableApp function disables two factor auth using
+func (s *TwoFactorAuthServiceOp) DisableApp(token string) (resp *Response, err error) {
+	headers := map[string]string{"x-otp-token": token}
+	return s.client.DoRequestWithHeader("DELETE", headers, twoFactorAuthAppPath, nil, nil)
+}
+
+// DisableSms function disables two factor auth using
+func (s *TwoFactorAuthServiceOp) DisableSms(token string) (resp *Response, err error) {
+	headers := map[string]string{"x-otp-token": token}
+	return s.client.DoRequestWithHeader("DELETE", headers, twoFactorAuthSmsPath, nil, nil)
+}
+
+// ReceiveApp orders the auth service to issue an SMS token
+func (s *TwoFactorAuthServiceOp) ReceiveApp() (otpURI string, resp *Response, err error) {
```

Comment:
I would rename this function to SeedApp, becuase the API call doesn't return a token but init info for 2FA App item - OTP type, secret seed and issuer.

With the results of the function, you're supposed to init your 2FA app, This has to be done only before you call the EnableApp. Then supply tokens from it.

Reply:
I'm a little late to this conversation as the code has been merged, but I disagree with using the function name SeedApp. "Seed" in this context does not make sense to me and does not follow any of the Packet API terminology. I think this will likely cause confusion for developers. It would for me.

I usually refer to the providers docs (usually API docs) if I need further clarification on what a function does, but in this case, I would never find any reference to "seed" and it's behavior will not be obvious.

I would recommend following the documentation terminology whenever possible.

Reply:
@edevenport thanks for your input. My opinion:

I suggested SeedApp because ReceiveApp does not make sense in the case to me. I see it semantically perfect that the `/user/otp/sms/receive` returns OTP for the API. The return value from `/user/otp/app/receive` is fundamentally different, and the call is used in completely different context.

There is no documentation to follow for the url `/user/otp/app/receive`. I had to check out what the method actually returns (it's otpauth:// URL). Nothing about it to to be found at https://www.packet.net/developers/api/twofactorauth/.

The otpauth returned from the POST to `/user/otp/app/receive` contains the secret key to the time-based hash in your 2FA app, e.g. Google Authenticator. It does not return a OTP which you can immediately use, like the `/user/otp/sms/receive` method. I see that as "seeding" the App mode of 2FA. "ReceiveApp" makes it sound that you will get an OTP from the API for the 2FA app which is nonsense - the OTP with the App is time-based hash.